tlatools/org.lamport.tlatools/src/tlc2/debug/TLCBreakpointExpression.java

```diff
@@ -0,0 +1,222 @@
+	 */
+	public static OpDefNode process(
+			final SpecProcessor processor,
+			final ModuleNode semanticRoot,
+			final Location location,
+			final String conditionExpr
+	) {
+		if (null == conditionExpr || conditionExpr.isBlank()) {
+			return null;
+		}
+		ToolIO.out.println("BPExpr: processing expression \"" + conditionExpr + "\"");
+
+		final String rootModName = semanticRoot.getName().toString();
+		final String bpModName = generateUnusedName(semanticRoot, "__BreakpointModule__%s");
+		ToolIO.out.println("BPExpr: wrapping with module \"" + bpModName + "\"");
+		final String bpOpName = generateUnusedName(semanticRoot, "__BreakpointExpr__%s");
+		final List<String> paramNames = getScopedIdentifiers(semanticRoot, location);
+		final String params = paramNames.size() > 0 ? "(" + String.join(", ", paramNames) + ")" : "";
+		final String bpOpDef = bpOpName + params;
+		ToolIO.out.println("BPExpr: wrapping with op \"" + bpOpDef + "\"");
+
+		final String wrapper = "---- MODULE %s ----\nEXTENDS %s\n%s == %s\n====";
+		String wrappedConditionExpr = String.format(wrapper, bpModName, rootModName, bpOpDef, conditionExpr);
+		byte[] wrappedConditionExprBytes = wrappedConditionExpr.getBytes(StandardCharsets.UTF_8);
+		InputStream sourceCode = new ByteArrayInputStream(wrappedConditionExprBytes);
+		TLAplusParser parser = new TLAplusParser(sourceCode, StandardCharsets.UTF_8.name());
+		boolean syntaxParseSuccess = parser.parse();
+		SyntaxTreeNode syntaxRoot = parser.ParseTree;
+		if (!syntaxParseSuccess || null == syntaxRoot) {
+			// Parse error is output to ToolIO.out
+			return null;
+		}
+		ToolIO.out.println("BPExpr: success [syntactic analysis]");
+
+		Errors semanticLog = new Errors();
+		SemanticNode.setError(semanticLog); // Annoyingly static
+		Generator semanticChecker = new Generator(semanticRoot.semanticChecker.moduleTable, semanticLog);
+		ModuleNode bpModule = null;
+		try {
+			bpModule = semanticChecker.generate(syntaxRoot);
+		} catch (AbortException e) {
+			ToolIO.err.print(e.toString());
+			return null;
+		}
+		if (null == bpModule || semanticLog.isFailure()) {
+			ToolIO.err.print(semanticLog.toString());
+			return null;
+		}
+		ToolIO.out.println("BPExpr: success [semantic analysis]");
+
+		// Run level-checking. The operator should be restricted to
+		// action-level or below.
+		Errors levelCheckingErrors = new Errors();
+		boolean levelCheckingSuccess = bpModule.levelCheck(levelCheckingErrors);
+		if (!levelCheckingSuccess || levelCheckingErrors.isFailure() || !bpModule.levelCorrect) {
+			ToolIO.err.println(levelCheckingErrors.toString());
+			return null;
+		}
+		ToolIO.out.println("BPExpr: success [level analysis]");
+		
+		OpDefNode bpOp = bpModule.getOpDef(bpOpName);
+		if (null == bpOp) {
+			ToolIO.err.println("ERROR: unable to find breakpoint expression op " + bpOpName);
+		}
+
+		if (!(LevelConstants.ConstantLevel == bpOp.level || LevelConstants.VariableLevel == bpOp.level
```

Comment:
This level check is sufficient for evaluating *Watch* expressions (not breakpoints). A Watch expression is evaluated for the currently active stack frame, regardless of whether that frame matches the expression’s level. If the evaluation fails, the debugger displays an appropriate message indicating that the Watch expression failed to evaluate.

Reply:
Does that mean I shouldn't implement the
>Conditionally restrict operator action level depending on breakpoint location

feature?

Reply:
Definitely not for Watch expressions, but I could imagine that novice users would appreciate it if the debugger would notify them if a breakpoint expression's level exceeds the scope's level. That said, does DAP even provide a mechanism to notify the user beyond visual breakpoint validation/verification (hollow vs. red breakpoint)?

Reply:
Not that I was able to find, other than hollow breakpoints + debug log.